documentation/TraceEvent/TraceEventProgrammersGuide.md

```diff
@@ -936,3 +936,41 @@ Unfortunately when a provider registers with the operating system (the Win32 `Ev
 ## On the fly filtering AND Logging using `ETWReloggerTraceEventSource`
 
 TODO
+
+## `UNMANAGED_CODE_TIME` and `CPU_TIME` when analyzing CPU traces collected with EventPipe
+
+EventPipe is a technology in .NET Core 3.1 on that allows the collection of events and CPU sampling on all platforms. The CPU sampling performed by EventPipe is only aware of managed code, which means that transitions into native code, e.g., P/Invokes, won't appear in the trace. This means that stacks that containing native code will end with the last managed frame. For example:
+
+```csharp
+namespace Example
+{
+    class Program
+    {
+        [DLLImport("MyNativeLibrary.so")]
+        public static extern void MyNativeFunction();
+
+        static void A() => B();
+        static void B() => C();
+        static void C() => MyNativeFunction();
+
+
+        static void main()
+        {
+           while (true)
+           {
+                DateTime then = DateTime.Now;
+                while (DateTime.Now - then < TimeSpan.FromSeconds(1)) {}
+                A();
+           }
+        }
+    }
+}
+```
+
+The above code will generate CPU samples on EventPipe that look like `Main` or `Main->A->B->C->MyNativeFunction` but contain none of the native stack that happens after `MyNativeFunction`.
+
+During a typical CPU usage investigation, you may want to ask about on-CPU vs off-CPU time to determine when your code is blocked waiting or actively doing work. Due to the limitation of only knowing about managed frames, CPU samples collected via EventPipe can't give exact on/off-CPU information. Instead, the TraceEvent library uses a heuristic to add pseudo-frames to the trave that indicate whether there are additional native frames on the stack or no. It uses relative time deltas between events and the state of the Garbage Collector at the time the sample was take to insert either `UNMANAGED_CODE_TIME` or `CPU_TIME` onto the stack in the "Thread Time" view or when exporting to SpeedScope or other formats. `UNMANAGED_CODE_TIME` represents stacks where there are one or more native frames after the last managed frame and that these frames may be blocked waiting _or_ actively on the CPU. `CPU_TIME` represents stacks where the last managed frame is the function currently on the CPU and doing work.
```

Comment:
The last two sentences seem to contradict the 3rd from last.

> `UNMANAGED_CODE_TIME` represents stacks where there are one or more native frames after the last managed frame and that these frames may be blocked waiting _or_ actively on the CPU. `CPU_TIME` represents stacks where the last managed frame is the function currently on the CPU and doing work.

I read these sentences saying that what matters is whether there are native frames at the leaf of the stack.

> It uses relative time deltas between events and the state of the Garbage Collector at the time the sample was take to insert either `UNMANAGED_CODE_TIME` or `CPU_TIME` onto the stack in the "Thread Time" view or when exporting to SpeedScope or other formats

I read this sentence saying that what matters is the GC state and time deltas.

Reply:
The phrasing is definitely confusing. The GC state is what informs the computer whether there were native leaf frames, and the relative time deltas tell us how long that was the case.

For example, if I have a function that is 50% managed work and 50% P/Invoked work, I'd end up with two sets of thread time stacks. One that is `MyFunction->CPU_TIME` corresponding to the aggregate of samples where the GC was in COOP mode and another that is `MyFunction->UNMANAGED_CODE_TIME` for when the GC was in PRE mode.

The duration of the `CPU_TIME` and `UNMANAGED_CODE_TIME` frames come from the relative deltas and which frame gets used comes from the GC mode during the sample.

Let me try to reword this so it's more obvious that all these things are describing the same set of information.

Reply:
I'd suggest deleting the part about GC state and time deltas in that case. I think the critical information for the reader is to understand what callstack produces what descriptive term. The techniques we use to determine the callstack are implementation details that appear unnecessary as long as they are accomplishing the job reliably.